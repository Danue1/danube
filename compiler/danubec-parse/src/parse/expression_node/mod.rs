mod literal_expression_node;

use crate::State;

impl crate::Context {
    pub fn expression_node(&mut self) -> State {
        one_of!(self.expression_literal_node())
    }
}
