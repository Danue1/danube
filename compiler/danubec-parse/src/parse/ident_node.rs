use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn ident_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, IDENT_NODE);
        self.finish_node();
        State::Stop
    }
}
