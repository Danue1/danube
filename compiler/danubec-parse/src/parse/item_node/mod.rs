mod enum_item_node;
mod struct_item_node;

impl crate::Context {
    pub fn item_node(&mut self) -> crate::State {
        one_of!(self.struct_item_node(), self.enum_item_node())
    }
}
