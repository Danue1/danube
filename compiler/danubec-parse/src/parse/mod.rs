mod ident_node;
mod item_node;
mod path_node;
mod type_node;

use crate::context::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::context::Context {
    #[inline]
    pub fn ast(&mut self) {
        self.start_node(SyntaxKind::AST_NODE);
        self.skip_whitespace();
        while !self.is_empty() {
            if self.item_node() == State::Continue {
                self.unexpected_token();
            }
            self.skip_whitespace();
        }
        self.finish_node();
    }

    fn unexpected_token(&mut self) {
        self.start_node(SyntaxKind::ERROR);
        self.bump();
        self.finish_node();
    }
}
