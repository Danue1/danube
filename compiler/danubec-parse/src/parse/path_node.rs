use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn path_node(&mut self) -> State {
        guard!(self, IDENT_KEYWORD, PATH_NODE, ident_node);

        while !self.is_eof() {
            if expect!(self, SyntaxKind::COLON) {
                if expect!(self, SyntaxKind::COLON) {
                    self.skip_whitespace();
                    self.ident_node();
                } else {
                    self.unexpected_token();
                }
            } else {
                break;
            }
        }

        self.finish_node();
        State::Stop
    }
}
