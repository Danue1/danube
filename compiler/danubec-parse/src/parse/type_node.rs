use crate::State;
use danubec_syntax_kind::SyntaxKind;

impl crate::Context {
    pub fn type_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.unnamed_type_kind_node() == State::Continue
            && self.path_type_kind_node() == State::Continue
        {
            State::Continue
        } else {
            self.start_node_at(checkpoint, SyntaxKind::TYPE_NODE);
            self.finish_node();
            State::Stop
        }
    }

    // ()
    // (type1)
    // (type1, type2)
    pub fn unnamed_type_kind_node(&mut self) -> State {
        guard!(self, LEFT_PAREN, UNNAMED_TYPE_KIND_NODE);
        self.skip_whitespace();

        while !self.is_eof() {
            if expect!(self, SyntaxKind::RIGHT_PAREN) {
                break;
            }

            let checkpoint = self.checkpoint();
            if self.type_node() == State::Continue {
                self.unexpected_token();
            } else {
                self.start_node_at(checkpoint, SyntaxKind::UNNAMED_TYPE_KIND_NODE);
                self.finish_node();
            }
            self.skip_whitespace();
            if expect!(self, SyntaxKind::COMMA) {
                self.skip_whitespace();
            }
        }

        self.finish_node();
        State::Stop
    }

    // Foo
    // Foo::Bar
    pub fn path_type_kind_node(&mut self) -> State {
        let checkpoint = self.checkpoint();
        if self.path_node() == State::Continue {
            return State::Continue;
        }

        self.start_node_at(checkpoint, SyntaxKind::PATH_TYPE_KIND_NODE);
        self.finish_node();
        State::Stop
    }
}
