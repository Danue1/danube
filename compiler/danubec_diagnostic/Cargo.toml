[package]
name = "danubec_diagnostic"
version = "0.1.0"
edition = "2024"

[dependencies]
