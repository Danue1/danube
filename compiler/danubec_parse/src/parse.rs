use crate::{
    event::Event,
    grammar::{Context, krate},
};
use danubec_diagnostic::Diagnostic;
use danubec_lex::lex;
use danubec_syntax::{GreenNodeBuilder, SyntaxKind, SyntaxNode};

pub fn parse(source: &str, diagnostic: &mut Diagnostic) -> SyntaxNode {
    let mut parse = |tokens: &[_]| {
        let tokens: Vec<_> = tokens.iter().map(|&(kind, _)| kind).collect();
        let mut context = Context::new(&tokens, diagnostic);
        krate(&mut context);

        context.finish()
    };

    let tokens = lex(source);
    let events = parse(&tokens);

    build(&tokens, events)
}

fn build(mut tokens: &[(SyntaxKind, &str)], mut events: Vec<Event>) -> SyntaxNode {
    let mut builder = GreenNodeBuilder::new();

    macro_rules! advance {
        () => {
            if let Some(&(kind, text)) = tokens.get(0) {
                builder.token(kind.into(), text);
                tokens = &tokens[1..];
            }
        };
    }

    macro_rules! trivia {
        () => {
            loop {
                match tokens.get(0) {
                    Some((kind, _)) if kind.at_trivia() => advance!(),
                    _ => break,
                }
            }
        };
    }

    for index in 0..events.len() {
        match std::mem::replace(&mut events[index], Event::Placeholder) {
            event @ Event::Start { forward_parent, .. }
            | event @ Event::Expire { forward_parent } => {
                let mut forward_parents = match event {
                    Event::Start { kind, .. } => vec![kind],
                    _ => vec![],
                };
                let mut index = index;
                let mut forward_parent = forward_parent;

                while let Some(fp) = forward_parent {
                    index += fp;
                    match std::mem::replace(&mut events[index], Event::Placeholder) {
                        Event::Start {
                            kind,
                            forward_parent: fp,
                        } => {
                            forward_parents.push(kind);
                            forward_parent = fp;
                        }
                        Event::Expire { forward_parent: fp } => {
                            forward_parent = fp;
                        }
                        _ => {
                            //
                        }
                    }
                }

                for kind in forward_parents.into_iter().rev() {
                    builder.start_node(kind.into());
                }

                trivia!();
            }
            Event::Token => {
                advance!();
                trivia!();
            }
            Event::End => builder.finish_node(),
            Event::Placeholder => {
                //
            }
        }
    }

    SyntaxNode::new_root(builder.finish())
}
