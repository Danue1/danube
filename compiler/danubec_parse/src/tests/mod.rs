use crate::parse;
use danubec_diagnostic::Diagnostic;

#[test]
fn top_level_attribute() {
    let source = r#"
#![a("value")]
#![a(key)]
#![a(key1, key2)]
#![a(key = "value")]
#![a::b(key = "value")]"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn attribute() {
    let source = r#"
#[a("value")]
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn zero_sized_struct() {
    let source = r#"
struct Foo;"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn named_struct() {
    let source = r#"
struct Foo { }
struct Bar {
    a: i32,
    b: String,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn unnamed_struct() {
    let source = r#"
struct Foo();
struct Bar(());
struct Baz(i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn mixed_enum() {
    let source = r#"
enum Foo {}
enum Bar {
    A,
    B { a: i32, b: String },
    C(i32, String),
    D = 42,
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn module() {
    let source = r#"
mod foo;
mod bar {
    struct Foo;
    enum Bar {
        A,
        B { a: i32, b: String },
        C(i32, String),
        D = 42,
    }
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn constant() {
    let source = r#"
const FOO: i32 = 42;
const BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn static_variable() {
    let source = r#"
static FOO: i32 = 42;
static BAR: str = "Hello, world!";"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn type_alias() {
    let source = r#"
type Foo = i32;
type Bar = (i32, String);"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn function() {
    let source = r#"
fn foo();
fn bar() {}
fn baz(a: i32, b: String) -> (i32, String) {
    (a, b)
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn trait_definition() {
    let source = r#"
trait Foo { }
trait Bar {
    fn foo();
    const BAR: i32;
    type Baz;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn anonymous_implement() {
    let source = r#"
impl Foo { }
impl Foo {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn targeted_implement() {
    let source = r#"
impl Foo for Bar { }
impl Foo for Bar {
    fn foo() {}
    const BAR: i32 = 42;
    type Baz = i32;
}"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}

#[test]
fn use_definition() {
    let source = r#"
use {};
use {{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use *;
use foo;
use foo as bar;
use foo::{};
use foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};

use ::{};
use ::{{},*,foo,bar as baz,::{},::*,::foo,::bar as baz,};
use ::*;
use ::foo;
use ::foo as bar;
use ::foo::{};
use ::foo::{*,foo,bar as baz,{},::*,::foo,::bar as baz,::{}};
"#;
    let diagnostic = Diagnostic::new();
    let (events, diagnostic) = parse(&source, diagnostic);

    dbg!(&diagnostic);

    assert!(diagnostic.is_empty());
    insta::assert_debug_snapshot!(events);
}
