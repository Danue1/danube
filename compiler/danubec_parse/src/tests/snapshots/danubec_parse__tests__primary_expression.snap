---
source: compiler/danubec_parse/src/tests/mod.rs
expression: node
---
ROOT_NODE@0..468
  NEW_LINE@0..1 "\n"
  DEFINITION_NODE@1..468
    FUNCTION_DEFINITION_NODE@1..468
      FN@1..3 "fn"
      WHITESPACE@3..4 " "
      IDENTIFIER_NODE@4..7
        IDENTIFIER_SEGMENT@4..7
          IDENTIFIER@4..7 "foo"
      LEFT_PAREN@7..8 "("
      RIGHT_PAREN@8..9 ")"
      WHITESPACE@9..10 " "
      FUNCTION_BODY_BLOCK_NODE@10..468
        BLOCK_EXPRESSION_NODE@10..468
          LEFT_BRACE@10..11 "{"
          NEW_LINE@11..12 "\n"
          WHITESPACE@12..16 "    "
          EXPRESSION_STATEMENT_NODE@16..27
            BREAK_EXPRESSION_NODE@16..21
              BREAK@16..21 "break"
            SEMICOLON@21..22 ";"
            NEW_LINE@22..23 "\n"
            WHITESPACE@23..27 "    "
          EXPRESSION_STATEMENT_NODE@27..41
            CONTINUE_EXPRESSION_NODE@27..35
              CONTINUE@27..35 "continue"
            SEMICOLON@35..36 ";"
            NEW_LINE@36..37 "\n"
            WHITESPACE@37..41 "    "
          EXPRESSION_STATEMENT_NODE@41..53
            RETURN_EXPRESSION_NODE@41..47
              RETURN@41..47 "return"
            SEMICOLON@47..48 ";"
            NEW_LINE@48..49 "\n"
            WHITESPACE@49..53 "    "
          EXPRESSION_STATEMENT_NODE@53..74
            FOR_EXPRESSION_NODE@53..74
              FOR@53..56 "for"
              WHITESPACE@56..57 " "
              PATH_PATTERN_NODE@57..59
                PATH_NODE@57..59
                  PATH_SEGMENT_IDENTIFIER_NODE@57..59
                    IDENTIFIER_NODE@57..59
                      IDENTIFIER_SEGMENT@57..59
                        IDENTIFIER@57..58 "i"
                        WHITESPACE@58..59 " "
              IN@59..61 "in"
              WHITESPACE@61..62 " "
              PATH_EXPRESSION_NODE@62..67
                PATH_NODE@62..67
                  PATH_SEGMENT_IDENTIFIER_NODE@62..67
                    IDENTIFIER_NODE@62..67
                      IDENTIFIER_SEGMENT@62..67
                        IDENTIFIER@62..66 "expr"
                        WHITESPACE@66..67 " "
              BLOCK_EXPRESSION_NODE@67..74
                LEFT_BRACE@67..68 "{"
                RIGHT_BRACE@68..69 "}"
                NEW_LINE@69..70 "\n"
                WHITESPACE@70..74 "    "
          EXPRESSION_STATEMENT_NODE@74..92
            WHILE_EXPRESSION_NODE@74..92
              WHILE@74..79 "while"
              WHITESPACE@79..80 " "
              LITERAL_EXPRESSION_NODE@80..85
                BOOLEAN_LITERAL_NODE@80..85
                  TRUE@80..84 "true"
                  WHITESPACE@84..85 " "
              BLOCK_EXPRESSION_NODE@85..92
                LEFT_BRACE@85..86 "{"
                RIGHT_BRACE@86..87 "}"
                NEW_LINE@87..88 "\n"
                WHITESPACE@88..92 "    "
          EXPRESSION_STATEMENT_NODE@92..104
            LOOP_EXPRESSION_NODE@92..104
              LOOP@92..96 "loop"
              WHITESPACE@96..97 " "
              BLOCK_EXPRESSION_NODE@97..104
                LEFT_BRACE@97..98 "{"
                RIGHT_BRACE@98..99 "}"
                NEW_LINE@99..100 "\n"
                WHITESPACE@100..104 "    "
          EXPRESSION_STATEMENT_NODE@104..119
            IF_EXPRESSION_NODE@104..119
              IF@104..106 "if"
              WHITESPACE@106..107 " "
              LITERAL_EXPRESSION_NODE@107..112
                BOOLEAN_LITERAL_NODE@107..112
                  TRUE@107..111 "true"
                  WHITESPACE@111..112 " "
              BLOCK_EXPRESSION_NODE@112..119
                LEFT_BRACE@112..113 "{"
                RIGHT_BRACE@113..114 "}"
                NEW_LINE@114..115 "\n"
                WHITESPACE@115..119 "    "
          EXPRESSION_STATEMENT_NODE@119..142
            IF_EXPRESSION_NODE@119..142
              IF@119..121 "if"
              WHITESPACE@121..122 " "
              LITERAL_EXPRESSION_NODE@122..127
                BOOLEAN_LITERAL_NODE@122..127
                  TRUE@122..126 "true"
                  WHITESPACE@126..127 " "
              BLOCK_EXPRESSION_NODE@127..130
                LEFT_BRACE@127..128 "{"
                RIGHT_BRACE@128..129 "}"
                WHITESPACE@129..130 " "
              ELSE@130..134 "else"
              WHITESPACE@134..135 " "
              BLOCK_EXPRESSION_NODE@135..142
                LEFT_BRACE@135..136 "{"
                RIGHT_BRACE@136..137 "}"
                NEW_LINE@137..138 "\n"
                WHITESPACE@138..142 "    "
          EXPRESSION_STATEMENT_NODE@142..241
            MATCH_EXPRESSION_NODE@142..241
              MATCH@142..147 "match"
              WHITESPACE@147..148 " "
              LITERAL_EXPRESSION_NODE@148..151
                INTEGER_LITERAL_NODE@148..151
                  INTEGER_SEGMENT@148..150 "42"
                  WHITESPACE@150..151 " "
              LEFT_BRACE@151..152 "{"
              NEW_LINE@152..153 "\n"
              WHITESPACE@153..161 "        "
              MATCH_ARM_NODE@161..168
                LITERAL_PATTERN_NODE@161..163
                  LITERAL_EXPRESSION_NODE@161..163
                    INTEGER_LITERAL_NODE@161..163
                      INTEGER_SEGMENT@161..162 "0"
                      WHITESPACE@162..163 " "
                EQUAL@163..164 "="
                RIGHT_CHEVRON@164..165 ">"
                WHITESPACE@165..166 " "
                BLOCK_EXPRESSION_NODE@166..168
                  LEFT_BRACE@166..167 "{"
                  RIGHT_BRACE@167..168 "}"
              COMMA@168..169 ","
              NEW_LINE@169..170 "\n"
              WHITESPACE@170..178 "        "
              MATCH_ARM_NODE@178..189
                OR_PATTERN_NODE@178..184
                  LITERAL_PATTERN_NODE@178..180
                    LITERAL_EXPRESSION_NODE@178..180
                      INTEGER_LITERAL_NODE@178..180
                        INTEGER_SEGMENT@178..179 "1"
                        WHITESPACE@179..180 " "
                  PIPE@180..181 "|"
                  WHITESPACE@181..182 " "
                  LITERAL_PATTERN_NODE@182..184
                    LITERAL_EXPRESSION_NODE@182..184
                      INTEGER_LITERAL_NODE@182..184
                        INTEGER_SEGMENT@182..183 "2"
                        WHITESPACE@183..184 " "
                EQUAL@184..185 "="
                RIGHT_CHEVRON@185..186 ">"
                WHITESPACE@186..187 " "
                BLOCK_EXPRESSION_NODE@187..189
                  LEFT_BRACE@187..188 "{"
                  RIGHT_BRACE@188..189 "}"
              COMMA@189..190 ","
              NEW_LINE@190..191 "\n"
              WHITESPACE@191..199 "        "
              MATCH_ARM_NODE@199..212
                AT_PATTERN_NODE@199..207
                  IDENTIFIER_NODE@199..203
                    IDENTIFIER_SEGMENT@199..203
                      IDENTIFIER@199..202 "foo"
                      WHITESPACE@202..203 " "
                  AT@203..204 "@"
                  WHITESPACE@204..205 " "
                  LITERAL_PATTERN_NODE@205..207
                    LITERAL_EXPRESSION_NODE@205..207
                      INTEGER_LITERAL_NODE@205..207
                        INTEGER_SEGMENT@205..206 "3"
                        WHITESPACE@206..207 " "
                EQUAL@207..208 "="
                RIGHT_CHEVRON@208..209 ">"
                WHITESPACE@209..210 " "
                BLOCK_EXPRESSION_NODE@210..212
                  LEFT_BRACE@210..211 "{"
                  RIGHT_BRACE@211..212 "}"
              COMMA@212..213 ","
              NEW_LINE@213..214 "\n"
              WHITESPACE@214..222 "        "
              MATCH_ARM_NODE@222..229
                PLACEHOLDER_PATTERN_NODE@222..224
                  PLACEHOLDER@222..223 "_"
                  WHITESPACE@223..224 " "
                EQUAL@224..225 "="
                RIGHT_CHEVRON@225..226 ">"
                WHITESPACE@226..227 " "
                BLOCK_EXPRESSION_NODE@227..229
                  LEFT_BRACE@227..228 "{"
                  RIGHT_BRACE@228..229 "}"
              COMMA@229..230 ","
              NEW_LINE@230..231 "\n"
              WHITESPACE@231..235 "    "
              RIGHT_BRACE@235..236 "}"
              NEW_LINE@236..237 "\n"
              WHITESPACE@237..241 "    "
          LET_STATEMENT_NODE@241..252
            LET@241..244 "let"
            WHITESPACE@244..245 " "
            PATH_PATTERN_NODE@245..246
              PATH_NODE@245..246
                PATH_SEGMENT_IDENTIFIER_NODE@245..246
                  IDENTIFIER_NODE@245..246
                    IDENTIFIER_SEGMENT@245..246
                      IDENTIFIER@245..246 "x"
            SEMICOLON@246..247 ";"
            NEW_LINE@247..248 "\n"
            WHITESPACE@248..252 "    "
          LET_STATEMENT_NODE@252..268
            LET@252..255 "let"
            WHITESPACE@255..256 " "
            PATH_PATTERN_NODE@256..257
              PATH_NODE@256..257
                PATH_SEGMENT_IDENTIFIER_NODE@256..257
                  IDENTIFIER_NODE@256..257
                    IDENTIFIER_SEGMENT@256..257
                      IDENTIFIER@256..257 "x"
            COLON@257..258 ":"
            WHITESPACE@258..259 " "
            PATH_TYPE_NODE@259..262
              PATH_NODE@259..262
                PATH_SEGMENT_IDENTIFIER_NODE@259..262
                  IDENTIFIER_NODE@259..262
                    IDENTIFIER_SEGMENT@259..262
                      IDENTIFIER@259..262 "i32"
            SEMICOLON@262..263 ";"
            NEW_LINE@263..264 "\n"
            WHITESPACE@264..268 "    "
          LET_STATEMENT_NODE@268..284
            LET@268..271 "let"
            WHITESPACE@271..272 " "
            PATH_PATTERN_NODE@272..274
              PATH_NODE@272..274
                PATH_SEGMENT_IDENTIFIER_NODE@272..274
                  IDENTIFIER_NODE@272..274
                    IDENTIFIER_SEGMENT@272..274
                      IDENTIFIER@272..273 "x"
                      WHITESPACE@273..274 " "
            EQUAL@274..275 "="
            WHITESPACE@275..276 " "
            LITERAL_EXPRESSION_NODE@276..278
              INTEGER_LITERAL_NODE@276..278
                INTEGER_SEGMENT@276..278 "42"
            SEMICOLON@278..279 ";"
            NEW_LINE@279..280 "\n"
            WHITESPACE@280..284 "    "
          LET_STATEMENT_NODE@284..305
            LET@284..287 "let"
            WHITESPACE@287..288 " "
            PATH_PATTERN_NODE@288..289
              PATH_NODE@288..289
                PATH_SEGMENT_IDENTIFIER_NODE@288..289
                  IDENTIFIER_NODE@288..289
                    IDENTIFIER_SEGMENT@288..289
                      IDENTIFIER@288..289 "x"
            COLON@289..290 ":"
            WHITESPACE@290..291 " "
            PATH_TYPE_NODE@291..295
              PATH_NODE@291..295
                PATH_SEGMENT_IDENTIFIER_NODE@291..295
                  IDENTIFIER_NODE@291..295
                    IDENTIFIER_SEGMENT@291..295
                      IDENTIFIER@291..294 "i32"
                      WHITESPACE@294..295 " "
            EQUAL@295..296 "="
            WHITESPACE@296..297 " "
            LITERAL_EXPRESSION_NODE@297..299
              INTEGER_LITERAL_NODE@297..299
                INTEGER_SEGMENT@297..299 "42"
            SEMICOLON@299..300 ";"
            NEW_LINE@300..301 "\n"
            WHITESPACE@301..305 "    "
          EXPRESSION_STATEMENT_NODE@305..320
            ARRAY_EXPRESSION_NODE@305..314
              LEFT_BRACKET@305..306 "["
              LITERAL_EXPRESSION_NODE@306..307
                INTEGER_LITERAL_NODE@306..307
                  INTEGER_SEGMENT@306..307 "0"
              COMMA@307..308 ","
              WHITESPACE@308..309 " "
              LITERAL_EXPRESSION_NODE@309..310
                INTEGER_LITERAL_NODE@309..310
                  INTEGER_SEGMENT@309..310 "1"
              COMMA@310..311 ","
              WHITESPACE@311..312 " "
              LITERAL_EXPRESSION_NODE@312..313
                INTEGER_LITERAL_NODE@312..313
                  INTEGER_SEGMENT@312..313 "2"
              RIGHT_BRACKET@313..314 "]"
            SEMICOLON@314..315 ";"
            NEW_LINE@315..316 "\n"
            WHITESPACE@316..320 "    "
          EXPRESSION_STATEMENT_NODE@320..345
            TUPLE_EXPRESSION_NODE@320..339
              LEFT_PAREN@320..321 "("
              LITERAL_EXPRESSION_NODE@321..323
                INTEGER_LITERAL_NODE@321..323
                  INTEGER_SEGMENT@321..323 "42"
              COMMA@323..324 ","
              WHITESPACE@324..325 " "
              LITERAL_EXPRESSION_NODE@325..332
                STRING_LITERAL_NODE@325..332
                  STRING_START@325..326 "\""
                  STRING_LITERAL_TEXT_NODE@326..331
                    STRING_SEGMENT@326..331 "hello"
                  STRING_END@331..332 "\""
              COMMA@332..333 ","
              WHITESPACE@333..334 " "
              LITERAL_EXPRESSION_NODE@334..338
                BOOLEAN_LITERAL_NODE@334..338
                  TRUE@334..338 "true"
              RIGHT_PAREN@338..339 ")"
            SEMICOLON@339..340 ";"
            NEW_LINE@340..341 "\n"
            WHITESPACE@341..345 "    "
          EXPRESSION_STATEMENT_NODE@345..357
            BLOCK_EXPRESSION_NODE@345..351
              LEFT_BRACE@345..346 "{"
              WHITESPACE@346..347 " "
              EXPRESSION_STATEMENT_NODE@347..350
                LITERAL_EXPRESSION_NODE@347..350
                  INTEGER_LITERAL_NODE@347..350
                    INTEGER_SEGMENT@347..349 "42"
                    WHITESPACE@349..350 " "
              RIGHT_BRACE@350..351 "}"
            SEMICOLON@351..352 ";"
            NEW_LINE@352..353 "\n"
            WHITESPACE@353..357 "    "
          EXPRESSION_STATEMENT_NODE@357..368
            FUNCTION_CALL_EXPRESSION_NODE@357..362
              PATH_EXPRESSION_NODE@357..360
                PATH_NODE@357..360
                  PATH_SEGMENT_IDENTIFIER_NODE@357..360
                    IDENTIFIER_NODE@357..360
                      IDENTIFIER_SEGMENT@357..360
                        IDENTIFIER@357..360 "foo"
              LEFT_PAREN@360..361
                LEFT_PAREN@360..361 "("
              RIGHT_PAREN@361..362 ")"
            SEMICOLON@362..363 ";"
            NEW_LINE@363..364 "\n"
            WHITESPACE@364..368 "    "
          EXPRESSION_STATEMENT_NODE@368..381
            FIELD_EXPRESSION_NODE@368..375
              PATH_EXPRESSION_NODE@368..371
                PATH_NODE@368..371
                  PATH_SEGMENT_IDENTIFIER_NODE@368..371
                    IDENTIFIER_NODE@368..371
                      IDENTIFIER_SEGMENT@368..371
                        IDENTIFIER@368..371 "foo"
              DOT@371..372
                DOT@371..372 "."
              IDENTIFIER_NODE@372..375
                IDENTIFIER_SEGMENT@372..375
                  IDENTIFIER@372..375 "bar"
            SEMICOLON@375..376 ";"
            NEW_LINE@376..377 "\n"
            WHITESPACE@377..381 "    "
          EXPRESSION_STATEMENT_NODE@381..396
            METHOD_CALL_EXPRESSION_NODE@381..390
              PATH_EXPRESSION_NODE@381..384
                PATH_NODE@381..384
                  PATH_SEGMENT_IDENTIFIER_NODE@381..384
                    IDENTIFIER_NODE@381..384
                      IDENTIFIER_SEGMENT@381..384
                        IDENTIFIER@381..384 "foo"
              DOT@384..385
                DOT@384..385 "."
              IDENTIFIER_NODE@385..388
                IDENTIFIER_SEGMENT@385..388
                  IDENTIFIER@385..388 "bar"
              LEFT_PAREN@388..389 "("
              RIGHT_PAREN@389..390 ")"
            SEMICOLON@390..391 ";"
            NEW_LINE@391..392 "\n"
            WHITESPACE@392..396 "    "
          EXPRESSION_STATEMENT_NODE@396..413
            FIELD_EXPRESSION_NODE@396..407
              FIELD_EXPRESSION_NODE@396..403
                PATH_EXPRESSION_NODE@396..399
                  PATH_NODE@396..399
                    PATH_SEGMENT_IDENTIFIER_NODE@396..399
                      IDENTIFIER_NODE@396..399
                        IDENTIFIER_SEGMENT@396..399
                          IDENTIFIER@396..399 "foo"
                DOT@399..400
                  DOT@399..400 "."
                IDENTIFIER_NODE@400..403
                  IDENTIFIER_SEGMENT@400..403
                    IDENTIFIER@400..403 "bar"
              DOT@403..404
                DOT@403..404 "."
              IDENTIFIER_NODE@404..407
                IDENTIFIER_SEGMENT@404..407
                  IDENTIFIER@404..407 "baz"
            SEMICOLON@407..408 ";"
            NEW_LINE@408..409 "\n"
            WHITESPACE@409..413 "    "
          EXPRESSION_STATEMENT_NODE@413..434
            METHOD_CALL_EXPRESSION_NODE@413..428
              METHOD_CALL_EXPRESSION_NODE@413..422
                PATH_EXPRESSION_NODE@413..416
                  PATH_NODE@413..416
                    PATH_SEGMENT_IDENTIFIER_NODE@413..416
                      IDENTIFIER_NODE@413..416
                        IDENTIFIER_SEGMENT@413..416
                          IDENTIFIER@413..416 "foo"
                DOT@416..417
                  DOT@416..417 "."
                IDENTIFIER_NODE@417..420
                  IDENTIFIER_SEGMENT@417..420
                    IDENTIFIER@417..420 "bar"
                LEFT_PAREN@420..421 "("
                RIGHT_PAREN@421..422 ")"
              DOT@422..423
                DOT@422..423 "."
              IDENTIFIER_NODE@423..426
                IDENTIFIER_SEGMENT@423..426
                  IDENTIFIER@423..426 "baz"
              LEFT_PAREN@426..427 "("
              RIGHT_PAREN@427..428 ")"
            SEMICOLON@428..429 ";"
            NEW_LINE@429..430 "\n"
            WHITESPACE@430..434 "    "
          EXPRESSION_STATEMENT_NODE@434..446
            INDEX_EXPRESSION_NODE@434..440
              PATH_EXPRESSION_NODE@434..437
                PATH_NODE@434..437
                  PATH_SEGMENT_IDENTIFIER_NODE@434..437
                    IDENTIFIER_NODE@434..437
                      IDENTIFIER_SEGMENT@434..437
                        IDENTIFIER@434..437 "foo"
              LEFT_BRACKET@437..438
                LEFT_BRACKET@437..438 "["
              LITERAL_EXPRESSION_NODE@438..439
                INTEGER_LITERAL_NODE@438..439
                  INTEGER_SEGMENT@438..439 "0"
              RIGHT_BRACKET@439..440 "]"
            SEMICOLON@440..441 ";"
            NEW_LINE@441..442 "\n"
            WHITESPACE@442..446 "    "
          EXPRESSION_STATEMENT_NODE@446..460
            ASSIGNMENT_EXPRESSION_NODE@446..454
              PATH_EXPRESSION_NODE@446..450
                PATH_NODE@446..450
                  PATH_SEGMENT_IDENTIFIER_NODE@446..450
                    IDENTIFIER_NODE@446..450
                      IDENTIFIER_SEGMENT@446..450
                        IDENTIFIER@446..449 "foo"
                        WHITESPACE@449..450 " "
              ASSIGNMENT_OPERATOR_NODE@450..453
                PLUS__EQUAL@450..453
                  PLUS@450..451 "+"
                  EQUAL@451..452 "="
                  WHITESPACE@452..453 " "
              LITERAL_EXPRESSION_NODE@453..454
                INTEGER_LITERAL_NODE@453..454
                  INTEGER_SEGMENT@453..454 "1"
            SEMICOLON@454..455 ";"
            NEW_LINE@455..456 "\n"
            WHITESPACE@456..460 "    "
          EXPRESSION_STATEMENT_NODE@460..467
            BINARY_EXPRESSION_NODE@460..465
              LITERAL_EXPRESSION_NODE@460..462
                INTEGER_LITERAL_NODE@460..462
                  INTEGER_SEGMENT@460..461 "1"
                  WHITESPACE@461..462 " "
              BINARY_OPERATOR_NODE@462..464
                PLUS@462..464
                  PLUS@462..463 "+"
                  WHITESPACE@463..464 " "
              LITERAL_EXPRESSION_NODE@464..465
                INTEGER_LITERAL_NODE@464..465
                  INTEGER_SEGMENT@464..465 "2"
            SEMICOLON@465..466 ";"
            NEW_LINE@466..467 "\n"
          RIGHT_BRACE@467..468 "}"
